kong/tools/string.lua

```diff
@@ -20,18 +22,185 @@ local CR_BYTE    = byte("\r")
 local _M = {}
 
 
-_M.join = pl_stringx.join
+_M.join = require("pl.stringx").join
 
 
---- splits a string.
+--- splits a string (kept for backward compatibility, use splitn instead).
 -- just a placeholder to the penlight `pl.stringx.split` function
 -- @function split
-_M.split = pl_stringx.split
+_M.split = require("pl.stringx").split
+
+
+local function split_once_common(value, pattern, plain)
+  if value == nil then
+    return nil, nil
+  elseif pattern == nil then
+    return value, nil
+  elseif pattern == "" then
+    return "", value
+  elseif value == "" then
+    return "", nil
+  end
+
+  local s, e = find(value, pattern, nil, plain)
+  if not s then
+    return value, nil
+  end
+
+  if s == 1 and e == 1 then
+    return "", (sub(value, e + 1))
+  end
+
+  return (sub(value, 1, s - 1)), (sub(value, e + 1))
+end
+
+
+--- splits a string once with a plain delimiter.
+-- @function split_once
+function _M.split_once(value, delim)
+  local k, v = split_once_common(value, delim, true)
+  return k, v
+end
```

Comment:
This should be the same, right? Can we do this?

```suggestion
function _M.split_once(value, delim)
  return split_once_common(value, delim, true)
end
```

Reply:
I benchmarked this and didn't see any difference. I can do that if that feels better to you.

Reply:
I have now pushed code that does tail calls. No benefit in perf, but may be worse in error messages.